lib/syscall_shim/uk_prsyscall.c

```diff
@@ -0,0 +1,709 @@
+	case SYS_write:
+		do {
+			int fd    = (int)    va_arg(args, long);
+			void *buf = (void *) va_arg(args, long);
+			__sz len  = (__sz)   va_arg(args, long);
+
+			PR_SYSCALL(sb, fmtf, syscall_num, PT_FD, fd,
+				   PT_BUFP(len), buf, PT_UDEC, len);
+			PR_SYSRET(sb, fmtf, PT_UDEC, rc);
+		} while (0);
+		break;
+	case SYS_read:
+		do {
+			int fd    = (int)    va_arg(args, long);
+			void *buf = (void *) va_arg(args, long);
+			__sz len  = (__sz)   va_arg(args, long);
+
+			PR_SYSCALL(sb, fmtf, syscall_num, PT_FD, fd,
+				   PT_BUFP(((rc >= 0) ? (__sz) rc : len))
+				   | ((rc >= 0) ? PT_OUT : 0x0), buf,
+				   PT_UDEC, len);
+			PR_SYSRET(sb, fmtf, PT_UDEC, rc);
+		} while (0);
+		break;
+#endif /* HAVE_uk_syscall_read && HAVE_uk_syscall_write */
+
+#ifdef HAVE_uk_syscall_stat
+	case SYS_stat:
+		VPR_SYSCALL(sb, fmtf, syscall_num, args,
+			    PT_PATH, PT_STRUCTSTAT);
+		PR_SYSRET(sb, fmtf, PT_STATUS, rc);
+		break;
+#endif /* HAVE_uk_syscall_stat */
+
+#ifdef HAVE_uk_syscall_fstat
+	case SYS_fstat:
+		VPR_SYSCALL(sb, fmtf, syscall_num, args, PT_FD, PT_STRUCTSTAT);
+		PR_SYSRET(sb, fmtf, PT_STATUS, rc);
+		break;
+#endif /* HAVE_uk_syscall_fstat */
+
+#ifdef HAVE_uk_syscall_close
+	case SYS_close:
+		VPR_SYSCALL(sb, fmtf, syscall_num, args, PT_FD);
+		PR_SYSRET(sb, fmtf, PT_STATUS, rc);
+		break;
+#endif /* HAVE_uk_syscall_close */
+
+#ifdef HAVE_uk_syscall_dup
+	case SYS_dup:
+		VPR_SYSCALL(sb, fmtf, syscall_num, args, PT_FD);
+		PR_SYSRET(sb, fmtf, PT_FD, rc);
+		break;
+#endif /* HAVE_uk_syscall_dup */
+
+#ifdef HAVE_uk_syscall_dup2
+	case SYS_dup2:
+		VPR_SYSCALL(sb, fmtf, syscall_num, args, PT_FD, PT_FD);
+		PR_SYSRET(sb, fmtf, PT_FD, rc);
+		break;
+#endif /* HAVE_uk_syscall_dup2 */
+
+#ifdef HAVE_uk_syscall_gettid
+	case SYS_gettid:
+		VPR_SYSCALL(sb, fmtf, syscall_num, args);
+		PR_SYSRET(sb, fmtf, PT_TID, rc);
+		break;
+#endif /* HAVE_uk_syscall_gettid */
+
+#ifdef HAVE_uk_syscall_getpid
+	case SYS_getpid:
+		VPR_SYSCALL(sb, fmtf, syscall_num, args);
+		PR_SYSRET(sb, fmtf, PT_PID, rc);
+		break;
+#endif /* HAVE_uk_syscall_gettid */
+
+	default:
+		do {
+			long arg0 = va_arg(args, long);
+			long arg1 = va_arg(args, long);
+
+			uk_streambuf_shcc(sb, fmtf, SYSCALL);
+			uk_streambuf_strcpy(sb, uk_syscall_name((syscall_num)));
+			uk_streambuf_shcc(sb, fmtf, RESET);
+			uk_streambuf_strcpy(sb, "(");
+			uk_streambuf_shcc(sb, fmtf, VALUE);
+			uk_streambuf_printf(sb, "0x%lx", arg0);
+			uk_streambuf_shcc(sb, fmtf, RESET);
+			uk_streambuf_strcpy(sb, ", ");
+			uk_streambuf_shcc(sb, fmtf, VALUE);
+			uk_streambuf_printf(sb, "0x%lx", arg1);
+			uk_streambuf_shcc(sb, fmtf, RESET);
+			uk_streambuf_strcpy(sb, ", ...)");
+			PR_SYSRET(sb, fmtf, PT_HEX, rc);
+		} while (0);
+		break;
+	}
+
+	/* Try to fix the line ending if content got truncated */
+	if (uk_streambuf_istruncated(sb)) {
+		/* reserve an extra byte for the newline ending if configured */
+		__sz needed_len = 3
+				  + (fmtf & UK_PRSYSCALL_FMTF_NEWLINE) ? 1 : 0;
```

Comment:
From the static analysis tool findings notes:

```
Clarify calculation precedence for '+' and '?'.
```

Reply:
Ouh, good point. I will add parentheses.